src/main/scala/io/circe/yaml/Parser.scala

```diff
@@ -29,37 +24,51 @@ final case class Parser(
     json <- yamlToJson(parsed)
   } yield json
 
-  def parse(yaml: String): Either[ParsingFailure, Json] = parse(new StringReader(yaml))
+  def parse(yaml: String): Either[ParsingFailure, Json] =
+    parse(new StringReader(yaml))
 
   def parseDocuments(yaml: Reader): Stream[Either[ParsingFailure, Json]] = parseStream(yaml).map(yamlToJson)
   def parseDocuments(yaml: String): Stream[Either[ParsingFailure, Json]] = parseDocuments(new StringReader(yaml))
 
-  private[this] def parseSingle(reader: Reader) =
-    Either.catchNonFatal(new Yaml(loaderOptions).compose(reader)).leftMap(err => ParsingFailure(err.getMessage, err))
+  private[this] def asScala[T](ot: Optional[T]): Option[T] =
+    if (ot.isPresent) Some(ot.get()) else None
+
+  private[this] def createComposer(reader: Reader) =
+    new Composer(settings, new ParserImpl(settings, new StreamReader(settings, reader)))
+
+  private[this] def parseSingle(reader: Reader): Either[ParsingFailure, Node] =
+    Either.catchNonFatal {
+      val composer = createComposer(reader)
+      asScala(composer.getSingleNode)
+    } match {
+      case Left(err)          => Left(ParsingFailure(err.getMessage, err))
+      case Right(None)        => Left(ParsingFailure("no document found", new RuntimeException("no document found")))
+      case Right(Some(value)) => Right(value)
+    }
 
   private[this] def parseStream(reader: Reader) =
-    new Yaml(loaderOptions).composeAll(reader).asScala.toStream
+    createComposer(reader).asScala.toStream
 
   private[this] object CustomTag {
-    def unapply(tag: Tag): Option[String] = if (!tag.startsWith(Tag.PREFIX))
+    def unapply(tag: Tag): Option[String] = if (!tag.getValue.startsWith(Tag.PREFIX))
       Some(tag.getValue)
     else
       None
   }
 
-  private[this] class FlatteningConstructor extends SafeConstructor {
+  private[this] class FlatteningConstructor(settings: LoadSettings) extends StandardConstructor(settings) {
     def flatten(node: MappingNode): MappingNode = {
       flattenMapping(node)
       node
     }
 
     def construct(node: ScalarNode): Object =
-      getConstructor(node).construct(node)
+      super.construct(node)
   }
 
   private[this] def yamlToJson(node: Node): Either[ParsingFailure, Json] = {
     // Isn't thread-safe internally, may hence not be shared
```

Comment:
Does this thread-safety comment still apply

Reply:
I don't know, to be honest. And since I don't know, I'm picking the conservative option of keeping things as they were.